c/tskit/genotypes.c

```diff
@@ -102,60 +102,76 @@ tsk_vargen_copy_alleles(tsk_vargen_t *self, const char **alleles)
     return ret;
 }
 
+static int
+init_samples_and_index_map(const tsk_treeseq_t *tree_sequence, const tsk_id_t *samples,
```

Comment:
I like the idea of abstracting this out, but I'm finding it hard to keep track of all the pointer dereferences (i.e., things like `` sizeof(**ret_sample_index_map)``). Perhaps it would be simpler to make this a private method of the vargen_t class, like ``vargen_init_samples_and_index_map(vargen_t *self, /* other params */)``? Then we wouldn't need to pass so many pointers around and it'd be a bit more obvious what the function is doing.

Reply:
That sounds better. I see many benefits to what you're suggesting. The benefit of the current function to being decoupled from all of self/tsk_vargen_t does not seem worth it.

It's interesting how despite `**ret_sample_index_map` being simpler than `*self->alt_sample_index_map` from the machine instruction/CPU perspective, the later is way easier for us to read and follow.

Reply:
Done. Here's the "historically preserved non-squashed" commit https://github.com/tskit-dev/tskit/commit/7cba3f1d9e30b34d136fecdb72fab48abf59f65e on branch [const_c_api](https://github.com/castedo/tskit/tree/const_c_api) that is now squashed into PR branch `pr/const_c_api` ... thanks to git-prepr :)